aggregator/src/aggregator/aggregation_job_creator.rs

```diff
@@ -347,6 +380,164 @@ impl<C: Clock + 'static> AggregationJobCreator<C> {
+        for<'a> &'a A::OutputShare: Into<Vec<u8>>,
+    {
+        Ok(self
+            .datastore
+            .run_tx(|tx| {
+                let (this, task) = (Arc::clone(&self), Arc::clone(&task));
+                Box::pin(async move {
+                    let min_batch_size = usize::try_from(task.min_batch_size())?;
+
+                    // Find unaggregated client reports & existing unfilled batches.
+                    let (mut unaggregated_report_ids, outstanding_batches) = try_join!(
+                        tx.get_unaggregated_client_report_ids_for_task(task.id()),
+                        tx.get_outstanding_batches_for_task(task.id())
+                    )?;
+
+                    // First attempt to allocate unaggregated reports to existing unfilled batches,
+                    // then generate new batches as necessary.
+                    let mut batch_iter = outstanding_batches
+                        .into_iter()
+                        .map(|outstanding_batch| (false, outstanding_batch))
+                        .chain(iter::repeat_with(|| {
+                            (true, OutstandingBatch::new(*task.id(), random(), 0, 0))
+                        }));
+
+                    // Main loop: repeatedly consume some of the unaggregated report IDs to generate
+                    // an aggregation job, assigning it to an existing batch which has need of
+                    // reports, or a new batch if no existing batch needs reports.
+                    let mut aggregation_jobs = Vec::<AggregationJob<L, FixedSize, A>>::new();
+                    let mut report_aggregations = Vec::<ReportAggregation<L, A>>::new();
+                    let mut new_batches = Vec::new();
+                    let (mut is_batch_new, mut batch) = batch_iter.next().unwrap();
+                    let mut max_size = batch.max_size();
+                    loop {
+                        // Figure out desired aggregation job size:
+                        //  * It can't be larger than the number of reports available.
+                        //  * It can't be larger than the configured maximum aggregation job size.
+                        //  * It can't be larger than the difference between the maximum batch size
+                        //    & the maximum number of reports that may end up aggregated into this
+                        //    batch based on already-existing aggregation jobs; otherwise, we risk
+                        //    aggregating more than max_batch_size reports together.
+                        // Choose the maximal size meeting all of these requirements.
+                        let aggregation_job_size = [
+                            unaggregated_report_ids.len(),
+                            this.max_aggregation_job_size,
+                            max_batch_size - max_size,
+                        ]
+                        .into_iter()
+                        .min()
+                        .unwrap();
```

Comment:
It'd be cool if there was some kind of `InfiniteIterator` trait for iterators that are guaranteed to always yield a value so that one could iterate without all the `unwrap()`s... but that's not a review comment on this change!

Reply:
👍🏻